pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,269 @@
+
+	// Generate unique user and bucket name
+	bucket := genUniqueBucketName(c.objContext.UID)
+	userConfig := c.genUserConfig()
+
+	// Create S3 user
+	logger.Debugf("creating s3 user object %q", userConfig.UserID)
+	user, rgwerr, err := CreateUser(c.objContext, userConfig)
+	if err != nil {
+		if rgwerr == ErrorCodeFileExists {
+			objectUser, _, err := GetUser(c.objContext, userConfig.UserID)
+			if err != nil {
+				err = errors.Wrapf(err, "failed to get details from ceph object user %q", objectUser.UserID)
+				updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+				return err
+			}
+
+			// Set access and secret key
+			s3AccessKey = *objectUser.AccessKey
+			s3SecretKey = *objectUser.SecretKey
+
+		} else {
+			err = errors.Wrapf(err, "failed to create s3 user %q. error code %d", userConfig.UserID, rgwerr)
+			updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+			return err
+		}
+	} else {
+		// Set access and secret key
+		s3AccessKey = *user.AccessKey
+		s3SecretKey = *user.SecretKey
+	}
+
+	// Initiate s3 agent
+	logger.Debug("initializing s3 connection")
+	s3client := CreateNewS3Client(s3endpoint, s3AccessKey, s3SecretKey)
+
+	// Force purge before starting anything
+	err = cleanupObjectHealthCheck(s3client, c.objContext.UID)
+	if err != nil {
+		err = errors.Wrap(err, "failed to perform object cleanup")
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Purge on exit
+	defer cleanupObjectHealthCheck(s3client, c.objContext.UID)
+
+	// Create S3 bucket
+	logger.Debugf("creating bucket %q", bucket)
+	_, err = s3client.CreateBucket(bucket)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to create bucket %q", bucket)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Put an object into the bucket
+	logger.Debugf("putting object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+	_, err = s3client.PutObjectInBucket(bucket, string(s3HealthCheckObjectBody), s3HealthCheckObjectKey, contentType)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to put object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Get the object from the bucket
+	logger.Debugf("getting object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+	read, err := s3client.GetObjectInBucket(bucket, s3HealthCheckObjectKey)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Compare the old and the existing object
+	logger.Debug("comparing objects md5")
+	oldMD5 := md5.Sum([]byte(s3HealthCheckObjectBody))
+	currentMD5 := md5.Sum([]byte(read))
+	if currentMD5 != oldMD5 {
```

Comment:
Using md5 for anything is discouraged from the security review (see #4572 for a specific case). How about sha256?

Reply:
Hum, it's a file hash I don't see any security implication.